internal/pkg/docker/dockerfile/dockerfile.go

```diff
@@ -89,31 +114,62 @@ func (df *Dockerfile) parse() error {
 	file, err := df.fs.Open(df.path)
 
 	if err != nil {
-		return fmt.Errorf("read dockerfile: %w", err)
+		return fmt.Errorf("read Dockerfile: %w", err)
 	}
 	defer file.Close()
 
-	scanner := bufio.NewScanner(file)
-	parsedDockerfile := parseFromScanner(scanner)
+	f, err := afero.ReadFile(df.fs, file.Name())
+	if err != nil {
+		return fmt.Errorf("read Dockerfile %s error: %w", f, err)
+	}
+	fileString := string(f)
+
+	parsedDockerfile, err := parseFromReader(fileString)
+	if err != nil {
+		return fmt.Errorf("parse Dockerfile: %w", err)
+	}
 
 	df.ExposedPorts = parsedDockerfile.ExposedPorts
+	df.HealthCheck = parsedDockerfile.HealthCheck
 	df.parsed = true
 	return nil
 }
 
-func parseFromScanner(scanner *bufio.Scanner) Dockerfile {
-	var line = ""
+func parseFromReader(line string) (Dockerfile, error) {
 	var df Dockerfile
 	df.ExposedPorts = []portConfig{}
-	for scanner.Scan() {
-		line = scanner.Text()
-		switch {
-		case strings.HasPrefix(line, "EXPOSE"):
-			currentPorts := parseExpose(line)
+	df.HealthCheck = healthCheck{}
+
+	// Parse(rwc io.Reader) reads lines from a Reader, and parses the lines into an AST.
+	ast, err := parser.Parse(strings.NewReader(line))
+	if err != nil {
+		return df, fmt.Errorf("parse reader: %w", err)
+	}
+
+	for _, child := range ast.AST.Children {
+		// ParseInstruction converts an AST to a typed instruction.
+		// Does prevalidation checks before parsing
+		// Example of an instruction is HEALTHCHECK CMD curl -f http://localhost/ || exit 1.
+		instruction, err := instructions.ParseInstruction(child)
+		if err != nil {
+			return df, fmt.Errorf("parse instructions: %w", err)
+		}
+		inst := fmt.Sprint(instruction)
+
+		// Getting the value at a children will return the Dockerfile directive
+		switch d := child.Value; d {
+		case "expose":
+			currentPorts := parseExpose(inst)
 			df.ExposedPorts = append(df.ExposedPorts, currentPorts...)
+		case "healthcheck":
+			healthCheckOptions, err := parseHealthCheck(inst)
```

Comment:
Are we importing `moby` because it allows us to give all the lines for an instruction as a single line?
Since we're doing our own parsing of the individual instructions the benefit seems to be pretty shallow 🤔 .

Reply:
It also does error validation and in the future if we want to parse other directives, this can help us easily parse the line and do whatever we want with it while again providing validations. I don't know how oddly formatted Dockerfiles can get but this simplifies the problem.